lib/Monads/trace/Trace_Prefix_Closed.thy

```diff
@@ -0,0 +1,301 @@
+  "prefix_closed A \<Longrightarrow> prefix_closed (notM A)"
+  unfolding notM_def
+  by (wpsimp wp: prefix_closed_terminal)
+
+lemma prefix_closed_ifM[prefix_closed_cond]:
+  "\<lbrakk> prefix_closed P; prefix_closed a; prefix_closed b \<rbrakk> \<Longrightarrow> prefix_closed (ifM P a b)"
+  unfolding ifM_def by wpsimp
+
+lemma prefix_closed_ifME[prefix_closed_cond]:
+  "\<lbrakk> prefix_closed P; prefix_closed a; prefix_closed b \<rbrakk> \<Longrightarrow> prefix_closed (ifME P a b)"
+  unfolding ifME_def by wpsimp
+
+lemma prefix_closed_whenM[prefix_closed_cond]:
+  "\<lbrakk> prefix_closed P; prefix_closed f \<rbrakk> \<Longrightarrow> prefix_closed (whenM P f)"
+  unfolding whenM_def
+  by (wpsimp wp: prefix_closed_terminal prefix_closed_cond)
+
+lemma prefix_closed_orM[prefix_closed_cond]:
+  "\<lbrakk> prefix_closed A; prefix_closed B \<rbrakk> \<Longrightarrow> prefix_closed (orM A B)"
+  unfolding orM_def
+  by (wpsimp wp: prefix_closed_terminal prefix_closed_cond)
+
+lemma prefix_closed_andM[prefix_closed_cond]:
+  "\<lbrakk> prefix_closed A; prefix_closed B \<rbrakk> \<Longrightarrow> prefix_closed (andM A B)"
+  unfolding andM_def
+  by (wpsimp wp: prefix_closed_terminal prefix_closed_cond)
+
+lemma prefix_closed_put_trace_elem[prefix_closed_terminal]:
+  "prefix_closed (put_trace_elem x)"
+  by (clarsimp simp: prefix_closed_def put_trace_elem_def)
+
+lemma prefix_closed_put_trace[prefix_closed_terminal]:
+  "prefix_closed (put_trace tr)"
+  by (induct tr; wpsimp wp: prefix_closed_terminal)
+
+lemma prefix_closed_interference[prefix_closed_terminal]:
+  "prefix_closed interference"
+  by (wpsimp simp: interference_def commit_step_def env_steps_def wp: prefix_closed_terminal)
+
+lemma prefix_closed_await[prefix_closed_terminal]:
+  "prefix_closed (Await c)"
+  by (wpsimp simp: Await_def wp: prefix_closed_terminal)
+
+lemma prefix_closed_repeat_n[prefix_closed_cond]:
+  "prefix_closed f \<Longrightarrow> prefix_closed (repeat_n n f)"
+  by (induct n; wpsimp wp: prefix_closed_terminal)
+
+lemma prefix_closed_repeat[prefix_closed_cond]:
+  "prefix_closed f \<Longrightarrow> prefix_closed (repeat f)"
+  apply (simp add: repeat_def)
+  apply (wpsimp wp: prefix_closed_terminal prefix_closed_cond)
+  done
+
+lemma prefix_closed_parallel[wp_split]:
+  "\<lbrakk>prefix_closed f; prefix_closed g\<rbrakk>
+   \<Longrightarrow> prefix_closed (parallel f g)"
+  apply (subst prefix_closed_def, clarsimp simp: parallel_def)
+  apply (subst (asm) zip.zip_Cons)
+  apply (clarsimp split: list.splits)
+  apply (drule(1) prefix_closedD)+
+  apply fastforce
+  done
+
+context begin
+(* We want to enforce that prefix_closed_terminal only contains lemmas that have no
+   assumptions. The following thm statement should fail if this is not true. *)
+private lemmas check_no_assumptions_internal = iffD1[OF refl, where P="prefix_closed f" for f]
+thm prefix_closed_terminal[atomized, THEN check_no_assumptions_internal]
+end
```

Comment:
Huh, I forgot that I added this check here. Are we ok with this approach, and if so should I add it to the other monad files that use the same attribute pattern?

It won't protect us from rules being incorrectly added to the sets past this point, but it should catch any mistakes we make adding new rules in this file.

Reply:
That's a neat idea! I wonder if there is a simpler version, but I can't come up with anything.

Happy to also add this in the other theories that could benefit from similar checks.

Reply:
To appease the thy linter, we could write `private lemmas check = ..` instead of `thm ..`

Reply:
If you're curious, this isn't completely new, I remembered that I had done something very similar in [rt](https://github.com/seL4/l4v/blob/rt/proof/refine/ARM/Invariants_H.thy#L3017).

For your second point, I think I have a very small preference for `thm ..` over `private lemmas check = ..`. It's pretty easy to ignore the thy linter failure here and it shouldn't reoccur in future PRs, as it feels very unlikely that these specific lines will need updating. On the other hand, it's minor but `private lemmas` aren't completely private. They can't actually be used anywhere else but they still pollute the namespace and can lead to mildly confusing `find_theorems` results.

Reply:
Ok, that is a valid point. Let's stick with `thm`, the linter won't be seeing this file very often.